src/main/java/org/apache/cassandra/distributed/api/LogAction.java

```diff
@@ -0,0 +1,352 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.distributed.api;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public interface LogAction
+{
+    /**
+     * @return current position of the iterator
+     * @see LogAction#grep(long, String)
+     * @see LogAction#watchFor(long, String)
+     */
+    long mark();
+
+    LineIterator matching(long startPosition, Predicate<String> fn);
+
+    default LineIterator matching(Predicate<String> fn){
+        return matching(Internal.DEFAULT_START_POSITION, fn);
+    }
+
+    default List<String> watchFor(long startPosition, Duration timeout, Predicate<String> fn) throws TimeoutException
```

Comment:
For what it is worth, the following APIs (watchFor variants) could be helpful as well. 

Within the timeout,
- expect such kind of message/pattern to appear for exact N times. 
- expect such kind of message/pattern to appear for at least N times. 
- expect no such message/pattern

Reply:
N times might make sense, but didn't see it in python-dtests so not sure if we need?

> expect no such message/pattern

```
Assertions.assertThat(logs.grep(...)).isEmpty();
```

think its not that useful in the API as it only lowers a bit of boilerplate

Reply:
Just thought of it. We can alway add it later when there is such usage.